packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java

```diff
@@ -65,138 +85,275 @@ private LocalAuthPlugin(Activity activity) {
+      activity.startActivityForResult(authIntent, LOCK_REQUEST_CODE);
+      return;
+    }
+
+    // Unable to authenticate
+    result.error("NotSupported", "This device does not support required security features", null);
+  }
+
+  private void authenticateSuccess() {
+    if (authInProgress.compareAndSet(true, false)) {
+      result.success(true);
+    }
+  }
+
+  private void authenticateFail() {
+    if (authInProgress.compareAndSet(true, false)) {
+      result.success(false);
+    }
+  }
+
+  /*
+   * Stops the authentication if in progress.
+   */
+  private void stopAuthentication() {
+    try {
+      if (authHelper != null && authInProgress.get()) {
+        authHelper.stopAuthentication();
+        authHelper = null;
       }
-    } else if (call.method.equals(("stopAuthentication"))) {
-      stopAuthentication(result);
-    } else {
-      result.notImplemented();
+      authInProgress.set(false);
+      result.success(true);
+    } catch (Exception e) {
+      result.success(false);
     }
   }
 
   /*
-   Stops the authentication if in progress.
-  */
-  private void stopAuthentication(Result result) {
+   * Returns biometric types available on device
+   */
+  private void getAvailableBiometrics(final Result result) {
     try {
-      if (authenticationHelper != null && authInProgress.get()) {
-        authenticationHelper.stopAuthentication();
-        authenticationHelper = null;
-        result.success(true);
+      Activity activity = getActivity();
+      if (activity == null || activity.isFinishing()) {
+        result.error("no_activity", "local_auth plugin requires a foreground activity", null);
         return;
       }
-      result.success(false);
+      ArrayList<String> biometrics = getAvailableBiometrics();
+      result.success(biometrics);
     } catch (Exception e) {
-      result.success(false);
+      result.error("no_biometrics_available", e.getMessage(), null);
+    }
+  }
+
+  private ArrayList<String> getAvailableBiometrics() {
+    ArrayList<String> biometrics = new ArrayList<String>();
+    Activity activity = getActivity();
+    if (activity == null || activity.isFinishing()) {
+      return biometrics;
     }
+    PackageManager packageManager = activity.getPackageManager();
+    if (Build.VERSION.SDK_INT >= 23) {
+      if (packageManager.hasSystemFeature(PackageManager.FEATURE_FINGERPRINT)) {
+        biometrics.add("fingerprint");
+      }
+    }
+    if (Build.VERSION.SDK_INT >= 29) {
+      if (packageManager.hasSystemFeature(PackageManager.FEATURE_FACE)) {
+        biometrics.add("face");
+      }
+      if (packageManager.hasSystemFeature(PackageManager.FEATURE_IRIS)) {
+        biometrics.add("iris");
+      }
+    }
+
+    return biometrics;
+  }
+
+  private boolean isDeviceSupported() {
+    KeyguardManager keyguardManager =
+        (KeyguardManager) getActivity().getBaseContext().getSystemService(KEYGUARD_SERVICE);
```

Comment:
If these services can be grabbed via ApplicationContext, prefer doing that and save  them as instance members to the plugin class. That way you don't have to get them with every call and eliminate the possibility that getActivity could be null.

Reply:
Removed `getActivity()`. Setting services and activity as instances variables when plugin registered and on attached to activity.